### PR TITLE
fixes #24918: add subregion dropdown to address editor

### LINKED ISSUE #24918: Add a State / Province dropdown in Address Editor
Replace the existing state input text field with a dropdown that displays a dropdown with the appropriate state / province values depending on the selected country. This should be done after https://github.com/mozilla-mobile/fenix/issues/24917

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20250)

### PR DESCRIPTION
**Pull request by MatthewTighe**
### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by MatthewTighe** [comment]
I also left a comment on the original issue detailing some of my findings for how desktop handles subregions.

File: app/src/main/res/layout/fragment_address_editor.xml
Comment (MozillaNoah) [inline app/src/main/res/layout/fragment_address_editor.xml:480]: follow-up: I don't believe the text buttons don't have the click ripple on them.

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:67]: One example is to checkout `CreditCardUtils`. There might be other examples of other files that are just managing a list or map of things. It probably makes sense to move the subregions and maybe country into their own file/object.

File: app/src/main/res/layout/fragment_address_editor.xml
Comment (gabrielluong) [inline app/src/main/res/layout/fragment_address_editor.xml:281]: I didn't build this yet, but wouldn't it make more sense for the subregion to appear before the zip?

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:228]: Can you add a reference to where you got the values of this list from in the code

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:34]: Are we posting the documentation anywhere using a tool like that wouldn't link the type next to the parameter name when it lists the function signature? Just curious why there's a precedent of including the type in the kDoc. The tools I've used to generate HTML pages for them before usually include the type right next to the parameter, and it's not a far stretch to access the type within the code itself. Mostly seems like an additional opportunity for the docs to get out of sync.

Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:34]: This question is above my pay grade lol. @gabrielluong do you have any insight to the history of our kdocs?

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:181]: I guess I was operating under the assumption that array adapters implemented some kind of recycler system, so that the position in a list wouldn't necessarily map to the position in the adapter. I'm trying to look up whether that's the case, but do you know for sure the lists never get out of sync?

Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:181]: There should still be a recycler system, but the selection index stuff should work as normal. My understanding is RecyclerViews have two running index systems (per your hesitation), but setting the selection index based on the underlying data collection will work as intended.

Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:181]: Oh yeah, I guess that was just a fundamental misunderstanding on my part of the purpose of array adapters. Thanks for pointing it out!

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:124]: I'm not sure I'm really understanding your suggestion. I structured it like this to try and keep the code as functional in style as possible and avoid the need for adding mutable `selectedCountryState` or inspecting the binding directly. I agree that having a `bind` function return a value is a little awkward, but `bindSubregionDropdown` does depend directly on the result of `bindCountryDropdown`. 

The other option is to just call `bindSubregionDropdown` from within `bindCountryDropdown`, but that makes `bindCountryDropdown` feel a little awkward as a name. Maybe just `bindDropdowns`?

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:34]: I don't think we are generating HTML pages for our docs. I didn't really think much about posting the function signature at the front here because it seemed to me it was being used as part of the sentence so I didn't really see anything wrong. If the intent was to post the literal function signature at the very beginning of everything, then I would probably say avoid that. When it comes to kDoc, I will typically defer to what I see is common in AC since that project has been richly documented due to the intent of project to be a public platform that would serve other projects to build browsers based on the libraries that are provided. Otherwise, I would prefer if we could adopt a style guide that is publicly available with appropriate linting to enforce. I don't know if there is such a thing for kDocs and Kotlin that is widely recognize across multiple organizations. An example would be something like https://google.github.io/styleguide/pyguide.html, https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html, https://prettier.io/

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:82]: parameter hints strike again. thanks!

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:124]: I was thinking just putting `bindCountryDropdown` into the root `bind` function and have a `val` for the parameter that `bindSubregionDropdown` needs. (to get ride of one function call to de-pastafy it)

semantically, I was just suggesting to put the contents of `bindCountryDropdown` into an `apply` block.

File: app/src/main/res/layout/fragment_address_editor.xml
Comment (MatthewTighe) [reply app/src/main/res/layout/fragment_address_editor.xml:281]: Oh yeah, it sure would. Good call

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:124]: Like moving the body of `bindCountryDropdown` into `bind`? It feels like quite a large function to me at that point, and I had also done this initial to try and more closely mirror what's in the [credit card editor](https://searchfox.org/mozilla-mobile/rev/5f5494ff3602921b88b0c1ba7263f4320d41494e/fenix/app/src/main/java/org/mozilla/fenix/settings/creditcards/view/CreditCardEditorView.kt#75). 

I'll push an update with the changes I made soon, but if you still end up not liking it I'm happy to make further edits

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:34]: Okay, I'll take the suggestions here. It sounds like, in general, there's no need to specify the `@param` type in its description in every case though?

**Comment by MatthewTighe** [comment]
Addressed first round of feedback. Now that I've looked at `CreditCardUtils` a bit, I think it would be pretty easy to refactor `Country` to be an enum class instead of a data class and would likely lead to cleaner code than all the map accesses. Didn't want to overstep on refactoring as part of this PR, but let me know what y'all think

File: app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt:34]: should this just be called "countryCode"?

File: app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt:45]: thoughts on making "CA" and "US" const vals?

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:98]: was the `selectedItem` call accidentally dropped?

**Review by MozillaNoah — COMMENTED** [review]
some quick comments while I wait for my local to build.

File: app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt:34]: This was intended to mirror the address reference data on desktop, but that does seem more clear for now.

File: app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/AddressUtils.kt:45]: I think it might make more sense to just change `Country` into an `enum class`, the country codes could effectively just be the names of subtypes in that case. I didn't want to overstep my bounds in refactoring as part of this last commit, but I'm happy to do that here or in a followup

File: app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/address/view/AddressEditorView.kt:98]: Oh definitely 😬 test should've definitely caught that as well, I'll make sure that's updated

**Review by MozillaNoah — APPROVED** [review]

**Comment by gabrielluong** [comment]
I don't think this needs another explicit review from me.

**Pull request merged**
